### ISSUE TITLE
Reinforced item shows wrong stats, negative damage?

### ISSUE THREAD

**Issue by Tharn** [CleverRaven/Cataclysm-DDA#19592]
Got a bit of a weird one for you here. I could even reproduce it.

What I saw the first time: 2 SCAR-H's. One had 90 dispersion, the other had 30! Both appeared to be ++reinforced. I then tried to repair both, which worked on the 30 dispersion gun and put it back at 90.

Anyway.. I loaded an earlier save and picked up the two SCAR-H's. I had to remove a bunch of mods from one of them. They were both at 90 dispersion. Then I saved and reloaded. They're now both at 30, still showing ++. Both able to be repaired. Could this be related to #19575?

http://www.filedropper.com/save_12

**Comment by Tharn** [comment]
After the JSON update, I still find many of my guns showing 'better than perfect' stats and able to be repaired down to a standard reinforced state. Is there a way to bring durability in line on old savegames or do I have to fix them all by hand?

**Comment by mugling** [comment]
Marking as confirmation required - can this be reproduced in a new save?

**Comment by Tharn** [comment]
I'd doubt it. At this point, I'm mostly looking to fix the save I have. A lot of damage values read like..

`"damage":1.666667e+000, "damage":1.010000e+000, "damage":3.400000e-001, "damage":3.337143e+000 `and so on...

Can I fix this by just going through all the map files with a replacer and setting, say, `"damage":1.0` everywhere? Or cutting the long values down to one number behind the comma? How do I properly convert it to the new system?

**Comment by mugling** [comment]
Yes, setting damage `1.0` should work. The bug arises because we output via stringstream but then read in via custom code and the two aren't commutative

**Comment by Tharn** [comment]
Cheers!

**Comment by Tharn** [comment]
Okay.. one last comment. I reset all of the scavenged items (indeed, the whole world present in .map files) back to `"damage":0`, then reinforced a bunch of weapons. In the file they're all listed as `"damage":-1.000000`, which I assume is correct.

However, they're still wonky. After a save>reload, a 90 dispersion gun turns once again into 30 dispersion, damage increases etc. And all reinforced guns are able to be repaired to regular reinforced status, which lasts only until you save>reload. I'll try this with a new game tomorrow.

**Comment by mugling** [comment]
What version?
Please post a save

**Comment by Tharn** [comment]
This is currently on Build 6007. Github won't let me upload zips so once again I'm using that crummy host.. sorry about that. http://www.filedropper.com/save_15

There's an A7 rifle in my inventory. It lists 26 damage, which is beyond what it should be able to do. You can 'reinforce' it down to 24. The other guns are in the trunk, bottom left tile.

**Comment by Tharn** [comment]
I just reproduced it with a new savegame, using a new world.

- Made a military character
- Edited a firearm repair kit into his inventory
- Accurized his starting weapon, an M1 Garand to a reinforced 64 damage, 90 dispersion.
- After a save>reload, the M1 Garand is at 66 damage, 30 dispersion.

Could you re-open this issue please?

**Comment by mugling** [comment]
> the only mods used are StatsThroughSkills, Disable NPC Needs, Disable Religious Texts & Prevent Zombie Revivification. Shouldn't matter for this I think.

Agreed

> Could you re-open this issue please?

Sure, can you post the save (preferably just copy-paste the part with the M1 Garand)

**Comment by Tharn** [comment]
The vanilla save without any firearm tampering shows:

` {
        "typeid": "garand",
        "item_vars": {
          "magazine_converted": "1"
        },
        "invlet": 99,
        "contents": [
          {
            "typeid": "garandclip",
            "contents": [
              {
                "typeid": "3006",
                "charges": 8
              }
            ]
          },
          {
            "typeid": "shoulder_strap"
          }
        ]
      },`

And then it doesn't seem to matter if I just accurize once and save, load the game once more so the gun shows faulty stats and save again, or if I 'fix' it and save again.. the save file always (correctly?) shows:

`{
        "typeid": "garand",
        "item_vars": {
          "magazine_converted": "1"
        },
        "invlet": 99,
        "damage": -1.000000,
        "contents": [
          {
            "typeid": "garandclip",
            "contents": [
              {
                "typeid": "3006",
                "charges": 8
              }
            ]
          },
          {
            "typeid": "shoulder_strap"
          }
        ]
      },`

**Comment by mugling** [comment]
So the save is `damage: -1` what does the display show?

**Comment by Tharn** [comment]
![faulty](https://cloud.githubusercontent.com/assets/13466852/20924764/91dd7fbe-bbb3-11e6-881b-8b5c5c59de43.png)

Shows this after loading the game.

![repaired](https://cloud.githubusercontent.com/assets/13466852/20924767/94abdeca-bbb3-11e6-9b4a-07269b9e85fc.png)

And this after applying the firearm repair kit.

**Comment by mugling** [comment]
So the bug is that you can apply the firearm kit despite the item already been reinforced?

**Comment by Tharn** [comment]
Yes, and that the gun shows incredibly good stats. 12 range, 62 damage and 90 dispersion are, as I understand it, the correct values for a reinforced Garand. But whenever I load a game with reinforced guns in it, they somehow get these stat buffs and become repairable, which removes the buffs until I reload the game.

**Comment by kevingranade** [comment]
What was the rationale for floating point damage values anyway?  They're
super fragile in weird ways.

**Comment by mugling** [comment]
It should be a continuous variable allowing deprecation of `if( x_in_y( 1, 10 ) { damage++; } )` constructs which players complained were too 'random'. I can see their point: why does one baseball bat breaks on the first swing but a different one lasts all season. We can now implement a lot better mechanics.

Display to the player remains stepped (to integers) and JSON definitions remain in integers. As an ancillary benefit we did get a lot of cleanup of the JSON loading code (which is scary-bad). Many of the bugs are due to problems in the repair code, many of which exist whether or not the underlying type is floating. Work is underway to correct this in #19663.

**Comment by mugling** [comment]
This bug isn't caused by floating point but a bug in the item repair code allowing double reinforcing

**Comment by mugling** [comment]
Also we only have 4 damage levels so a rescaling (by 100) would still require the same stepping logic. C++ floating point is well defined - the problem here is a bad conditional.

**Comment by Tharn** [comment]
Hey guys, since mugling's attempt at this was unfortunately rejected (see his PR), is there anything on the horizon in the way of a possible fix? AFAIK old issues are auto-closed and this is I think a fairly prominent one that all players experience when they play the game at one point or another.

**Comment by Coolthulhu** [comment]
It has been a while. Does it still reproduce?

**Comment by Tharn** [comment]
Yep, just tested it. Same thing as was described. I also noticed the M1 Garand seems to be gone from the game (aw ;)).

**Comment by Coolthulhu** [comment]
Can't reproduce. Did this:
* Spawn 2 SCAR-Hs
* Accurize both
* Save and load
* Try to accurize either
* Fail
* Damage one and re-accurize it
* Still can't over-accurize it
* Spawn a fresh one
* Save+load
* Can't over-accurize it

It could be just bugged saves. Does it reproduce with fresh items? If not, it only requires some save recovery code, which would be much easier than trying to locate the source of this weird behavior.

**Comment by Tharn** [comment]
Nope, still does it. Check my short procedure from further up the thread. Get yourself a kit and Mechanics level 10 for this. This allows you to reinforce the guns, and after a save-reload you see increased stats on a reinforced gun that you can then 'accurize away' by using the kit another time.

**Comment by ZenZen2** [comment]
Confirming this bug still exist.
Its almost like cheating, makes everyting that is reinforced stronger after reloading the game.

Additional info:
After the 'reinforced+reloadedGame' Item gets damaged, it show its real state after reloading the game.

**Comment by Qrox** [comment]
@BorkBorkGoesTheCode #24493 should allow guns to be fully reinforced, but I could not reproduce the bug described in this issue, with or without #24493. @Tharn @ZenZen2 could you please see if this issue still persists?

**Comment by Tharn** [comment]
One issue I still see: Sometimes guns will only reinforce up to bright green ++, which may turn into dark green ++ after a save -> reload.

**Comment by Qrox** [comment]
After #24493 partially-reinforced items (which show as bright green) should not appear again, and existing ones should be reinforce-able to the full condition. I don't see why reloading would cause them to turn into dark green however, somewhere in the save loading code that converts the double damage value into an integer perhaps?

**Comment by Tharn** [comment]
I think it has to do with them being damaged prior to being repaired. Will try to reproduce.

**Comment by Qrox** [comment]
After switching to MinGW-w64 this also happened to me, and I think I have found the cause of this issue: in `json.cpp`, the method `JsonIn::get_float()` uses `i * std::pow(10.0f, e + mod_e)` for float deserialization. This of course, depends heavily on the actual implementation of `std::pow()`. I could have proposed a fix, but unfortunately c++11 only comes with locale-specific float conversion functions (including `istream::operator>>()') which are slow, and not guaranteed to precisely restore serialized float values. c++17 do have a quick and precise conversion function, but that's out of the question for this project. I tried consecutively dividing/multiplying by 10 and it worked for this specific case, but I doubt that will not become broken under other circumstances or compilers.
